swift/extractor/main.cpp

```diff
@@ -3,15 +3,58 @@
 #include <stdlib.h>
 
 #include <swift/Basic/LLVMInitialize.h>
+#include <swift/FrontendTool/FrontendTool.h>
+
+#include "Extractor.h"
+
+using namespace std::string_literals;
+
+class Observer : public swift::FrontendObserver {
+ public:
+  explicit Observer(const codeql::Configuration& config) : config{config} {}
+
+  void performedSemanticAnalysis(swift::CompilerInstance& instance) override {
+    codeql::Extractor extractor{config, instance};
+    extractor.extract();
+  }
+
+ private:
+  const codeql::Configuration& config;
+};
+
+static std::string getenv_default(const char* envvar, const std::string& def) {
+  if (const char* var = getenv(envvar)) {
+    return var;
+  }
+  return def;
+}
 
 int main(int argc, char** argv) {
+  if (argc == 1) {
+    /// TODO: print usage
+    return 1;
+  }
+  /// The frontend can be called in different modes, we are only interested
+  /// in -frontend mode
+  /// TODO: filter out at the tracer level
+  if ("-frontend"s != argv[1]) {
+    return 0;
+  }
   PROGRAM_START(argc, argv);
-  if (auto trapDir = getenv("CODEQL_EXTRACTOR_SWIFT_TRAP_DIR")) {
-    std::string file = trapDir;
-    file += "/my_first.trap";
-    if (std::ofstream out{file}) {
-      out << "answer_to_life_the_universe_and_everything(42)\n";
+
+  codeql::Configuration configuration{};
```

Comment:
Optional: This function is getting quite long. I'd suggest factoring out some of the config setup. Some ideas:
 1. All-in-one: `codeql::Configuration swift_config_from_environment_and_cli_args(int argc, char** argv);`
 1a. As above, but as a static member `codeql::Configuration::from_environment_and_cli_args` for a less clunky name.
 2. Two parts: 
```
// Populates trapDir and sourceArchiveDir
codeql::Configuration swift_configuration_from_environment();
std::vector<std::string> swift_frontend_options_from_cli_args(int argc, char** argv);
```
 3. Just the second part (`swift_frontend_options_from_cli_args`), and keep the rest in line.
 
I think the all-in-one in particularly nice because we can pass the result directly to the `Observer` constructor (if it's going to make a copy). But it does have quite a clunky name unless you can come up with something better.

Reply:
It will grow even more soon, so I'd gather more code here before we can factor out some parts.

Reply:
I think that's all the more reason to refactor now, so when we try to add functionality here, understanding the flow doesn't slow us down!

Reply:
This function is ~30 LoC, I don't think it's worth refactoring it right now.